src/combinatorics/simple_graph/connectivity.lean

```diff
@@ -755,4 +788,115 @@ edges_bypass_subset _
+huv.elim (λ p, ⟨p.reverse⟩)
+
+@[trans] protected lemma reachable.trans {u v w : V}
+  (huv : G.reachable u v) (hvw : G.reachable v w) :
+  G.reachable u w :=
+huv.elim (λ puv, hvw.elim (λ pvw, ⟨puv.append pvw⟩))
+
+lemma reachable_iff_refl_trans_gen (u v : V) :
+  G.reachable u v ↔ relation.refl_trans_gen G.adj u v :=
+begin
+  split,
+  { rintro ⟨h⟩,
+    induction h,
+    { refl, },
+    { exact (relation.refl_trans_gen.single h_h).trans h_ih, }, },
+  { intro h,
+    induction h with _ _ _ ha hr,
+    { refl, },
+    { exact reachable.trans hr ⟨walk.cons ha walk.nil⟩, }, },
+end
+
+variables (G)
+
+lemma reachable_is_equivalence : equivalence G.reachable :=
+mk_equivalence _ (@reachable.refl _ G) (@reachable.symm _ G) (@reachable.trans _ G)
+
+/-- The equivalence relation on vertices given by `simple_graph.reachable`. -/
+def reachable_setoid : setoid V := setoid.mk _ G.reachable_is_equivalence
+
+/-- The connected components of a graph are elements of the quotient of vertices by
+the `simple_graph.reachable` relation. -/
+def connected_component := quot G.reachable
+
+/-- A graph is preconnected if every pair of vertices is reachable from one another. -/
+def preconnected : Prop := ∀ (u v : V), G.reachable u v
+
+/-- A graph is connected if it's preconnected and contains at least one vertex.
+This follows the convention observed by mathlib that something is connected iff it has
+exactly one connected component.
+
+There is a `has_coe_to_fun` instance so that `h u v` can be used instead
+of `h.preconnected u v`. -/
+@[protect_proj]
+structure connected : Prop :=
+(preconnected : G.preconnected)
+(nonempty : nonempty V)
+
+instance : has_coe_to_fun G.connected (λ _, Π (u v : V), G.reachable u v) :=
+⟨λ h, h.preconnected⟩
+
+/-- Gives the connected component containing a particular vertex. -/
+def connected_component_of (v : V) : G.connected_component := quot.mk G.reachable v
+
+instance connected_component.inhabited [inhabited V] : inhabited G.connected_component :=
+⟨G.connected_component_of default⟩
+
+lemma preconnected.subsingleton_connected_component (h : G.preconnected) :
+  subsingleton G.connected_component :=
+⟨λ c d, quot.ind (λ v d, quot.ind (λ w, quot.sound (h v w)) d) c d⟩
+
+/-- A subgraph is connected if it is connected as a simple graph. -/
+abbreviation subgraph.connected {G : simple_graph V} (H : G.subgraph) : Prop := H.coe.connected
+
+/-- A graph is *acyclic* (or a *forest*) if it has no cycles.
+A characterization: `simple_graph.is_acyclic_iff`.-/
+def is_acyclic : Prop := ∀ (v : V) (c : G.walk v v), ¬c.is_cycle
+
+/-- A *tree* is a connected acyclic graph. -/
+def is_tree : Prop := G.connected ∧ G.is_acyclic
+
+/-- A graph is *k-edge-connected* if it remains connected whenever
+fewer than k edges are removed. -/
+def edge_connected (k : ℕ) : Prop :=
+∀ (s ⊆ G.edge_set) (h : s.finite), h.to_finset.card < k → (G.delete_edges s).connected
```

Comment:
Perhaps it would be better to have these in a separate PR (so they can come with lemmas about them)? If not, I think they should be a different file so this one doesn't get too long

Reply:
Yeah, that makes sense.